src/coreclr/src/jit/morph.cpp

```diff
@@ -2890,9 +2890,72 @@ void Compiler::fgInitArgInfo(GenTreeCall* call)
     SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR structDesc;
 #endif // UNIX_AMD64_ABI
 
+#if defined(DEBUG)
+    // Check that we have valid inforamtion about call's argument types.
+    // For example:
+    // load byte; call(int) -> CALL(PUTARG_TYPE byte(IND byte));
+    // load int; call(byte) -> CALL(PUTARG_TYPE int (IND int));
+    // etc.
+    if ((call->callSig != nullptr) && (Target::g_tgtArgOrder == Target::ARG_ORDER_R2L))
+    {
+        GenTreeCall::Use* nodeArgs = call->gtCallArgs;
+
+        unsigned nodeArgsCount = 0;
+        for (GenTreeCall::Use* i = nodeArgs; i != nullptr; i = i->GetNext())
+        {
+            nodeArgsCount++;
+        }
+
+        CORINFO_SIG_INFO* sig          = call->callSig;
+        const unsigned    sigArgsCount = sig->numArgs;
+        assert(nodeArgsCount >= sigArgsCount);
+
+        // TODO: Delete this condition, it is a temporary solution to see if anything else fails.
+        // After tail call transformations we could have more nodes in the list + special arguments +
+        // setup trees. I need to find a clean way to match this list with the original signature if
+        // is possible. If not we will be left without a check that PUTARG_TYPE survive as they should.
+        if (nodeArgsCount == sigArgsCount)
```

Comment:
Should this be `!=`? Otherwise, the loop below will never execute, right?

Reply:
It is a temporary condition, the idea here is to match the signature with the call arguments as we have them in morph and check that we still have the correct types. However, it is tricky to do because we could have some nodes added to `gtCallArgs` between importation and call to this function (also this function can be called several times). 
So we need to skip all additional nodes that were added and I have problems with identifying them. I am planning to fix it once I pass all the other tests.